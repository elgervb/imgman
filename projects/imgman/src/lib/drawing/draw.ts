import { Brush } from './brushes/brush';
import { getMousePosition } from './utils';

export function enableDrawing(canvas: HTMLCanvasElement, brush: Brush, isDrawingEnabled: () => boolean) {
  let isDrawing = false;
  canvas.onmousedown = (evt: MouseEvent) => {
    if (isDrawingEnabled()) {
      isDrawing = true;

      brush.down(getMousePosition(evt, canvas));
    }
  };

  canvas.onmousemove = (evt: MouseEvent) => {
    if (isDrawingEnabled() && isDrawing) {
      brush.move(getMousePosition(evt, canvas));
    }
  };

  canvas.onmouseup = (evt: MouseEvent) => {
    if (isDrawingEnabled()) {
      isDrawing = false;
      brush.up(getMousePosition(evt, canvas));
    }
  };
}

export function disableDrawing(canvas: HTMLCanvasElement) {
  canvas.onmousedown = null;
  canvas.onmousemove = null;
  canvas.onmouseup = null;
}
