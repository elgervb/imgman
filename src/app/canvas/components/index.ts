export * from './canvas-image/canvas-image.component';
export * from './color-picker/color-picker.component';
