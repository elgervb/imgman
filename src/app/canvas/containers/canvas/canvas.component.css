
:host {
  align-items: center;
  display: flex;
  height: 100vh;
  justify-content: center;
  width: 100%;
}

app-toolbar {
  position: absolute;
  text-shadow: #fff 0px 0px 0.8rem;
  top: 1rem;
  width: 10rem;
}

.colorpicker-toolbar {
  left: 1rem;
}

.drawing-toolbar {
  left: 11.5rem;
}
.stencil-toolbar {
  left: 22rem;
}
