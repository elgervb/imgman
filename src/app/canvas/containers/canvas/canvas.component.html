
<app-canvas-image
  image="/assets/ugly-dog.png"
  (ready)="canvasIsReady($event)"
  appDraw
  [canvas]="canvas"
  [color]="color"
  [lineWidth]="drawingSettings?.lineWidth"
  [globalAlpha]="drawingSettings?.opacity"
  [brushType]="drawingSettings?.brushType"
  [drawingEnabled]="drawingSettings?.enabled">
</app-canvas-image>

<!-- TODO: create a color picker toolbar -->
<evb-flyout #flyout  header="Drawing" open="true">

  <ng-container *ngIf="canvas && colorWheel">
    <app-color-picker [canvas]="canvas" [secondaryCanvas]="colorWheel" (picked)="setColor($event)"></app-color-picker>
  </ng-container>
  <div>
    <app-canvas-image image="/assets/color-wheel.png" max="100" (ready)="colorWheelReady($event)"></app-canvas-image>
  </div>

  <app-drawing-toolbar (changed)="setDrawingSettings($event)"></app-drawing-toolbar>

</evb-flyout>

<div class="flyout-handle" (click)="showFlyout()"></div>
