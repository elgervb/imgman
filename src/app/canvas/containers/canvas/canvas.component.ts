import { Component, ElementRef, Input, OnInit, ViewChild } from '@angular/core';
import { PickedFile } from '@elgervb/stencil-components/dist/types/components/file-picker/pickedfile';
import { Rgb } from 'src/app/imgman/color/rgb';

import { DrawingSettings } from '../../components/drawing-toolbar/drawing-toolbar.component';

@Component({
  selector: 'app-canvas',
  templateUrl: './canvas.component.html',
  styleUrls: ['./canvas.component.css']
})
export class CanvasComponent implements OnInit {

  @Input() image: string;

  constructor() { }

  @ViewChild('flyout') flyout: ElementRef<HTMLEvbFlyoutElement>;

  canvas: HTMLCanvasElement;

  // TODO: move this to a color picker toolbar
  colorWheel: HTMLCanvasElement;
  color: Rgb;
  colorPickerActive: boolean;

  drawingSettings: DrawingSettings;

  progress: number;

  ngOnInit() {
    //
  }

  canvasIsReady(canvas: HTMLCanvasElement) {
    this.canvas = canvas;
  }

  colorWheelReady(colorWheel: HTMLCanvasElement) {
    this.colorWheel = colorWheel;
  }

  setColor(color: Rgb) {
    setTimeout(() => this.color = color, 0);
  }

  setImage(event: CustomEvent<PickedFile>) {
    this.image = event.detail.dataUrl;
  }

  setDrawingSettings(drawingSettings: DrawingSettings) {
    this.drawingSettings = drawingSettings;
  }

  stencilClick(event: UIEvent) {
    console.log('Got a click from a StencilJS button', event);
  }

  stencilCompleted(event: Event) {
    console.log('Got a completed event from a StencilJS progressbar', event);
  }

  setProgress(progress: number) {
    this.progress = progress;
  }

  setColorPickerActive(isActive: boolean) {
    this.colorPickerActive = isActive;
  }

  showFlyout() {
    this.flyout.nativeElement.toggle(true);
  }
}
