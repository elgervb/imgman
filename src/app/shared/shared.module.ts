import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { ToolbarComponent } from './components/toolbar/toolbar.component';
import { EvbValueAccessorDirective } from './directives/evbcontrolvalueaccessor';
import { IsDarkPipe } from './pipes/is-dark.pipe';
import { RgbToStringPipe } from './pipes/rgb-to-string.pipe';
import { SanitizePipe } from './pipes/sanitize.pipe';

@NgModule({
  declarations: [
    EvbValueAccessorDirective,
    SanitizePipe,
    ToolbarComponent,
    IsDarkPipe,
    RgbToStringPipe,
  ],
  imports: [
    CommonModule
  ],
  exports: [
    EvbValueAccessorDirective,
    IsDarkPipe,
    RgbToStringPipe,
    SanitizePipe,
    ToolbarComponent
  ]
})
export class SharedModule { }
