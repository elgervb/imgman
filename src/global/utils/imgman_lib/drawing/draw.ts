import { Brush } from './brushes/brush';
import { getMousePosition } from './utils';

export function enableDrawing(canvas: HTMLCanvasElement, brush: Brush, isDrawingEnabled: () => boolean) {
  let isDrawing = false;
  canvas.onmousedown = (evt: MouseEvent) => {
    if (isDrawingEnabled()) {
      isDrawing = true;

      brush.down(getMousePosition(evt, canvas));
    }
  };

  canvas.onmousemove = (evt: MouseEvent) => {
    if (isDrawingEnabled() && isDrawing) {
      brush.move(getMousePosition(evt, canvas));
    }
  };

  canvas.onmouseup = (evt: MouseEvent) => {
    if (isDrawingEnabled()) {
      isDrawing = false;
      brush.up(getMousePosition(evt, canvas));
    }
  };
}

export function disableDrawing(canvas: HTMLCanvasElement) {
  canvas.onmousedown = undefined;
  canvas.onmousemove = undefined;
  canvas.onmouseup = undefined;
}
