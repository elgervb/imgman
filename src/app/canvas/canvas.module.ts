import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';

import { SharedModule } from '../shared/shared.module';

import { CanvasRoutingModule } from './canvas-routing.module';
import { CanvasImageComponent, ColorPickerComponent } from './components';
import { DrawingToolbarComponent } from './components/drawing-toolbar/drawing-toolbar.component';
import { CanvasComponent } from './containers';
import { DrawDirective } from './directives/draw.directive';

@NgModule({
  declarations: [
    CanvasComponent,
    ColorPickerComponent,
    CanvasImageComponent,
    DrawDirective,
    DrawingToolbarComponent
  ],
  imports: [
    CommonModule,
    CanvasRoutingModule,
    ReactiveFormsModule,
    SharedModule
  ]
})
export class CanvasModule { }
