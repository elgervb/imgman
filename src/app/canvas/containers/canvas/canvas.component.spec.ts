import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { SharedModule } from 'src/app/shared/shared.module';

import { CanvasImageComponent, ColorPickerComponent } from '../../components';
import { DrawDirective } from '../../directives/draw.directive';

import { CanvasComponent } from './canvas.component';

describe('CanvasComponent', () => {
  let component: CanvasComponent;
  let fixture: ComponentFixture<CanvasComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        CanvasComponent,
        ColorPickerComponent,
        CanvasImageComponent,
        DrawDirective
      ],
      imports: [SharedModule]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CanvasComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
