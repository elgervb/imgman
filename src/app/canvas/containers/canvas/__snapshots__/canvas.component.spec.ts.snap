// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasComponent should create 1`] = `
<app-canvas
  flyout="undefined"
>
  <evb-dropzone>
    <evb-filepicker
      accept="image/*"
    >
      <h3
        class="pick-image-title"
      >
        Pick an image to get started
      </h3>
      <evb-button
        pill="true"
      >
        Pick a image
      </evb-button>
    </evb-filepicker>
  </evb-dropzone>
</app-canvas>
`;
